File: benchmarks/tsb/bench_to_numeric_dispatch.ts
/**
 * Benchmark: toNumeric generic dispatcher — exported toNumeric(value) dispatches to Series/array/scalar paths.
 * Mirrors pandas pd.to_numeric() with multiple input types.
 * Outputs JSON: {"function": "to_numeric_dispatch", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, toNumeric } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 30;

const strNums = Array.from({ length: SIZE }, (_, i) => String(i * 1.5));
const s = new Series({ data: strNums });

for (let i = 0; i < WARMUP; i++) {
  toNumeric(strNums, { errors: "coerce" });
  toNumeric(s, { errors: "coerce" });
  toNumeric("42.7");
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  toNumeric(strNums, { errors: "coerce" });
  toNumeric(s, { errors: "coerce" });
  toNumeric("42.7");
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "to_numeric_dispatch",
    mean_ms: Math.round((total / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_to_numeric_dispatch.py
"""Benchmark: toNumeric generic — pd.to_numeric() with array, Series, and scalar inputs."""
import json, time
import pandas as pd

SIZE = 50_000
WARMUP = 5
ITERATIONS = 30

str_nums = [str(i * 1.5) for i in range(SIZE)]
s = pd.Series(str_nums)

for _ in range(WARMUP):
    pd.to_numeric(str_nums, errors="coerce")
    pd.to_numeric(s, errors="coerce")
    pd.to_numeric("42.7")

start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.to_numeric(str_nums, errors="coerce")
    pd.to_numeric(s, errors="coerce")
    pd.to_numeric("42.7")
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "to_numeric_dispatch",
    "mean_ms": round(total / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total, 3),
}))
